teps/0060-remote-resource-resolution.md

```diff
@@ -319,19 +324,252 @@ Why should this TEP _not_ be implemented?
+  # ...
+```
+
+2. Custom Task reconciler is waiting for `gitref.tekton.dev/v1alpha1.RemoteFetch`. Fetches
+remote resource based on parameters from embedded spec.
+
+3. Custom Task reconciler creates `TaskRun` with fetched resource, passing through params,
+workspaces, etc...
+
+4. Custom Task reconciler mirrors status changes, results, etc from the `TaskRun` it created
+back to the `Run` that Pipeline reconciler is watching.
+
+Pros:
+- Reuses existing Custom Tasks pattern.
+- Doesn't require any changes in Pipelines or specification effort at all.
+- Lifetime of fetched resource can be tied to owner reference - the `TaskRun` that the custom task reconciler
+creates can add an owner reference to the `Run` for this I think?
+- Tekton _could_ provide one of these for git
+- Resource resolution is performed declaratively.
+- Resolution is executed concurrently while the Pipelines controller does other things.
+- Flexible since Custom Tasks can do "whatever they want".
+    - Example: if a Custom Task is implementing an -as-code flow it can fetch a Pipeline at a specific commit __AND__ re-write any referenced tasks in that Pipeline so that they too reference the specific commit as well.
+- Pipelines not responsible for anything Resolver-specific.
+- Development burden of writing a reconciler _could_ be alleviated with a custom task SDK?
+
+Cons:
+- No consistent specification of behaviour, potentially varies for every custom task implementing remote fetch.
+- No validation of fetched content by Pipelines.
+- No opportunity to apply defaults by Pipelines.
+- No opportunity for Pipelines reconciler to optimize fetching multiple remote refs up-front before a PipelineRun is allowed to start.
+- The custom task needs to have permissions to submit new taskruns to the cluster.
+- Unclear how we'd ever standardize anything around this.
+- Writing a reconciler could be considered difficult burden for users.
+- Weird split from user POV: Why does git reference use a completely different syntax from oci?
+- Pipelines doesn't distinguish between Custom Tasks and remote resolution, meaning it's much harder to surface metrics around latencies, errors, etc... that are specific to remote resolution.
+
+### Introduce a tekton resource request CRD
+
+Add a new CRD for async resource resolution. One process requests a resource by creating an object of this new kind. Another process fulfills the request by writing the resource's content directly inline to the object. The first process notices the updated condition and reads the resource from the object.
+
+1. User submits a run type with a remote reference:
+
+```yaml
+kind: TaskRun
+metadata:
+  name: foo-run
+spec:
+  taskRef:
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    remote:
+      apiVersion: gitref.tekton.dev/v1alpha1
+      url: https://github.com/tektoncd/catalog.git
+      ref: main
+      path: /tasks/git-clone/0.4/git-clone.yaml
+```
+
+2. Pipelines controller creates request object:
```

Comment:
This will pollute the cluster incredibly with `TektonResourceRequests`... I think using a CRD is good idea, but I would not associate the CRD to the specific request, rather to a resource provider which accepts parameters, similar to `ClusterInterceptors` in triggers.

Reply:
I'm not sure how/why we qualify one thing as pollution but another not though - why is one object (Task) not pollution but two objects (Task + ResourceRequest) pollution?  Are there specific object limits in Kubernetes that you have in mind here, or something else, that can help guide perspective on this?

Reply:
I've captured this in the cons section and added a couple of possible mitigations to the problem as well.

Reply:
this "pollution" would also apply to custom tasks (which create runs)